### ISSUE TITLE
Unnecessary recompilation

### ISSUE THREAD

**Issue by apt1002** [rrthomas/smite#234]
Sometimes I do "make check" and wait for specializer.c to compile, then I do "make install" and I wait again. Why is it recompiling?

**Comment by rrthomas** [comment]
I'm having trouble reproducing this (despite having seen what looks like unnecessary compilation quite a few times myself). I've tried "make clean" followed by "make" then "make install" at the top level, and also "make clean" followed by "make" in `lib` then `src` followed by "make install" at the top level (to see if cross-directory dependencies are in play). (I've also tried "make clean", "make check" then "make install".)

**Comment by apt1002** [comment]
Changing `gen-labels` seems to trigger this, IIRC.

**Comment by rrthomas** [comment]
@apt1002, I can't reproduce by:

    touch src/features/gen-labels
    make
    make check

That is, the `make check` does not remake any files already made by the preceding `make`.

**Comment by rrthomas** [comment]
I was able to reproduce this issue partially: `make` followed by `make check` unnecessarily recompiled `src/features/{auto-extend,predictor,core-dump}.c`. I was unable to work out why from looking at the dependencies.

The most obvious sort of reason would be that if one of the rules for C/Python code generation touched or replaced a file not mentioned as a dependency of the rule. This would be especially likely to cause problems in a parallel build (but not sure if you're using parallel make, @apt1002?).

I looked at the various manual dependencies I inserted between particular object and source files some time ago, and found they are all duplicated in the auto-generated dependencies files (under `src/.deps` and `src/features/.deps`), and some of the rules referred to files that no longer exist, so I removed them all (branch `fix-unnecessary-recompilation-234`).

I'd very much like any further help, in particular with reproducibly recompiling `specializer.c` unnecessarily.

**Comment by rrthomas** [comment]
Having removed the extra rules, the Travis and AppVeyor builds are broken: they all try to compile `storage.c` before building `src/include/mit/opcodes.h`, which is required via `mit.h`.

I don't understand this. I tried a fresh checkout without parallel Make on my machine, and it worked. (I am using GCC 7; Travis fails with both GCC 5.4 and GCC 8; I think AppVeyor on Ubuntu is using GCC 7, and not sure what it's using on Windows.)

For now I've reverted the change (except for removing references to files that no longer exist).
